File: Algorithms and Data Structures/iterative_binary_search.py
def binary_search(list, target):
    first = 0
    last = len(list) - 1

    while first <= last:
        midpoint = (first + last)//2

        if list[midpoint] == target:
            return midpoint
        elif list[midpoint] < target:
            first = midpoint + 1
        else:
            last = midpoint - 1
    return None


def verify(index):
    if index is not None:
        print("Target found at index: ", index)
    else:
        print("Target not found in list")


numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

result = binary_search(numbers, 6)
verify(result)

# The given Python code defines a function called binary_search that performs a binary search on a sorted list to find the index position of a target element. The function takes two arguments, a sorted list and a target element, and returns the index position of the target element if found in the list, else returns None.

# The binary_search function initializes two variables, first and last, to the indices of the first and last elements of the list, respectively. It then uses a while loop to repeatedly divide the search interval in half until the target element is found or the search interval is empty.

# At each iteration of the loop, the function calculates the midpoint of the search interval using the formula (first + last)//2. If the midpoint element of the list is equal to the target element, the function returns the index position of the midpoint. If the midpoint element is less than the target, the search is continued in the right half of the search interval by updating the first index to be midpoint + 1. If the midpoint element is greater than the target, the search is continued in the left half of the search interval by updating the last index to be midpoint - 1.

# If the target element is not found after iterating through the entire list, the function returns None.

# The code also defines another function called verify that takes an index position as an argument and prints a message indicating whether the target element was found at that index or not.

# Finally, the code creates a sorted list of numbers and calls the binary_search function with a target value of 6. The index position of the target value is stored in a variable called result, which is then passed to the verify function to print a message indicating whether the target element was found in the list or not.

File: Algorithms and Data Structures/Algorithms/iterative_binary_search/iterative_binary_search.js
function binarySearch(array, key) {
	let left = 0;
	let right = array.length - 1;
	while (left <= right) {
		const mid = left + Math.floor((right - left) / 2);
		if (array[mid] === key) {
			return mid;
		}
		if (array[mid] < key) {
			left = mid + 1;
		} else {
			right = mid - 1;
		}
	}
	return -1;
}

// The function takes two arguments: an array and a key value to search for.
// It then initializes two variables left and right that represent the leftmost and rightmost indices of the array.
// It then enters a while loop that continues as long as the left index is less than or equal to the right index.
// Inside the loop, it computes the midpoint index using the formula (left + right) / 2, and rounds down to the nearest integer using Math.floor().
// It then checks if the value at the midpoint index is equal to the target key. If it is, the function returns the midpoint index.
// If the value at the midpoint index is less than the target key, the function updates the left index to be one more than the midpoint, and continues searching the right half of the array.
// If the value at the midpoint index is greater than the target key, the function updates the right index to be one less than the midpoint, and continues searching the left half of the array.
// If the target value is not found in the array, the function returns -1.

// This function implements a binary search algorithm, which is a more efficient way to search sorted arrays compared to linear search. It works by repeatedly dividing the search interval in half, and discarding the half that cannot contain the target value until the target value is found, or the interval becomes empty.
